exonum/src/blockchain/service.rs

```diff
@@ -408,6 +409,25 @@ impl SharedNodeState {
                 .peers_info
                 .insert(c.addr(), *p);
         }
+
+        self.update_majority_count(state);
+    }
+
+    /// Returns the majority count from the current "State"
+    /// of a blockchain node.
+    pub fn majority_count(&self) -> usize {
+        self.state
+            .read()
+            .expect("Expected read lock.")
+            .majority_count
+    }
+
+    /// Updates internal majority count, from `State` of a blockchain node.
+    fn update_majority_count(&self, state: &State) {
```

Comment:
Now, since this function doesn't anymore have complicated logic, I suggest to remove it. Logic itself can be moved to the `update_node_state`, and `update_node_state` can be rewritten like this:

```
    pub fn update_node_state(&self, state: &State) {
        let mut lock = self.state.write().expect("Expected write lock.");

        lock.peers_info.clear();
        lock.majority_count = state.majority_count();

        for (p, c) in state.peers().iter() {
             lock.peers_info.insert(c.addr(), *p);
        }
    }
```

so the lock will be acquired only once.
@DarkEld3r do you approve?

Reply:
It is good idea.

Reply:
I am not sure that this is better. `write` on `RwLock` is fairly cheap and I don't think it is worth making whole method atomic. Besides, other methods in this `impl` block acquire the lock each time and I see no good reason for changing that tradition. (IMHO)

Reply:
There is one reason: it's more readable. And until benches show that performance became worse, I don't think that we should think about it.
"Premature optimization is the root of all evil".

And for the future I hope that we'll rewrite other cases too.
(IMHO :upside_down_face: )